{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collatz Conjecture\n",
    "\n",
    "The Collatz Conjecture or 3x+1 problem can be summarized as follows:\n",
    "\n",
    "Take any positive integer n. If n is even, divide n by 2 to get n / 2. If n is\n",
    "odd, multiply n by 3 and add 1 to get 3n + 1. Repeat the process indefinitely.\n",
    "The conjecture states that no matter which number you start with, you will\n",
    "always reach 1 eventually.\n",
    "\n",
    "Given a number n, return the number of steps required to reach 1.\n",
    "\n",
    "## Examples\n",
    "\n",
    "Starting with n = 12, the steps would be as follows:\n",
    "\n",
    "0. 12\n",
    "1. 6\n",
    "2. 3\n",
    "3. 10\n",
    "4. 5\n",
    "5. 16\n",
    "6. 8\n",
    "7. 4\n",
    "8. 2\n",
    "9. 1\n",
    "\n",
    "Resulting in 9 steps. So for input n = 12, the return value would be 9.\n",
    "\n",
    "## Notes\n",
    "\n",
    "The Collatz Conjecture is only concerned with strictly positive integers, so your solution should raise a `ValueError` with a meaningful message if given 0 or a negative integer.\n",
    "\n",
    "\n",
    "## Exception messages\n",
    "\n",
    "Sometimes it is necessary to raise an exception. When you do this, you should include a meaningful error message to\n",
    "indicate what the source of the error is. This makes your code more readable and helps significantly with debugging. Not\n",
    "every exercise will require you to raise an exception, but for those that do, the tests will only pass if you include\n",
    "a message.\n",
    "\n",
    "To raise a message with an exception, just write it as an argument to the exception type. For example, instead of\n",
    "`raise Exception`, you shold write:\n",
    "\n",
    "```python\n",
    "raise Exception(\"Meaningful message indicating the source of the error\")\n",
    "```\n",
    "\n",
    "## Submitting Exercises\n",
    "\n",
    "Note that, when trying to submit an exercise, make sure the solution is in the `$EXERCISM_WORKSPACE/python/collatz-conjecture` directory.\n",
    "\n",
    "You can find your Exercism workspace by running `exercism debug` and looking for the line that starts with `Workspace`.\n",
    "\n",
    "For more detailed information about running tests, code style and linting,\n",
    "please see the [help page](http://exercism.io/languages/python).\n",
    "\n",
    "## Source\n",
    "\n",
    "An unsolved problem in mathematics named after mathematician Lothar Collatz [https://en.wikipedia.org/wiki/3x_%2B_1_problem](https://en.wikipedia.org/wiki/3x_%2B_1_problem)\n",
    "\n",
    "## Submitting Incomplete Solutions\n",
    "It's possible to submit an incomplete solution so you can see how others have completed the exercise.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def collatz_steps(number):\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "\n",
    "\n",
    "\n",
    "# Tests adapted from `problem-specifications//canonical-data.json` @ v1.2.0\n",
    "\n",
    "class CollatzConjectureTests(unittest.TestCase):\n",
    "\n",
    "    def test_zero_steps_for_one(self):\n",
    "        self.assertEqual(collatz_steps(1), 0)\n",
    "\n",
    "    def test_divide_if_even(self):\n",
    "        self.assertEqual(collatz_steps(16), 4)\n",
    "\n",
    "    def test_even_and_odd_steps(self):\n",
    "        self.assertEqual(collatz_steps(12), 9)\n",
    "\n",
    "    def test_large_number_of_even_and_odd_steps(self):\n",
    "        self.assertEqual(collatz_steps(1000000), 152)\n",
    "\n",
    "    def test_zero_is_invalid_input(self):\n",
    "        with self.assertRaisesWithMessage(ValueError):\n",
    "            collatz_steps(0)\n",
    "\n",
    "    def test_negative_number_is_invalid_input(self):\n",
    "        with self.assertRaisesWithMessage(ValueError):\n",
    "            collatz_steps(-15)\n",
    "\n",
    "    # Utility functions\n",
    "    def setUp(self):\n",
    "        try:\n",
    "            self.assertRaisesRegex\n",
    "        except AttributeError:\n",
    "            self.assertRaisesRegex = self.assertRaisesRegexp\n",
    "\n",
    "    def assertRaisesWithMessage(self, exception):\n",
    "        return self.assertRaisesRegex(exception, r\".+\")\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
